lib/fs.js

```diff
@@ -1369,6 +1369,109 @@ function mkdirSync(path, options) {
   }
 }
 
+/*
+ * An recurssive algorithm for reading the entire contents of the `basePath` directory.
+ * This function does not validate `basePath` as a directory. It is passed directly to
+ * `binding.readdir`.
+ * @param {string} basePath
+ * @param {{ encoding: string, withFileTypes: boolean }} options
+ * @param {(
+ *   err?: Error,
+ *   files?: string[] | Buffer[] | Dirent[]
+ * ) => any} callback
+ * @returns {void}
+*/
+function readdirRecursive(basePath, options, callback) {
+  const context = {
+    withFileTypes: Boolean(options.withFileTypes),
+    encoding: options.encoding,
+    basePath,
+    readdirResults: [],
+    pathsQueue: [basePath],
+  };
+
+  let i = 0;
+
+  function read(path) {
+    const req = new FSReqCallback();
+    req.oncomplete = (err, result) => {
+      if (err) {
+        callback(err);
+        return;
+      }
+
+      if (result === undefined) {
+        callback(null, context.readdirResults);
+        return;
+      }
+
+      processReaddirResult({
+        result,
+        currentPath: path,
+        context,
+      });
+
+      if (i < context.pathsQueue.length) {
+        read(context.pathsQueue[i++]);
+      } else {
+        callback(null, context.readdirResults);
+      }
+    };
+
+    binding.readdir(
+      path,
+      context.encoding,
+      context.withFileTypes,
+      req,
+    );
+  }
+
+  read(context.pathsQueue[i++]);
+}
+
+function processReaddirResult({ result, currentPath, context }) {
+  // Calling `readdir` with `withFileTypes=true`, the result is an array of arrays.
+  // The first array is the names, and the second array is the types.
+  // They are guaranteed to be the same length; hence, setting `length` to the length
+  // of the first array within the result.
+  if (context.withFileTypes) {
+    handleDirents({ result, currentPath, context });
+  } else {
+    handleFilePaths({ result, currentPath, context });
+  }
+}
+
+function handleDirents({ result, currentPath, context }) {
```

Comment:
Would you mind adding jsdoc to all newly/updated functions?

Reply:
I haven't added jsdoc to these ones because they are not public, they are used inside `readdirRecursive` (which has JSDoc)